src/orange/orangeinp/detail/LogicUtils.cc

```diff
@@ -0,0 +1,220 @@
+            reading_surf = false;
+        }
+
+        // Parse a logic token
+        // NOLINTNEXTLINE(bugprone-switch-missing-default-case)
+        switch (v)
+        {
+                // clang-format off
+            case '*': result.push_back(logic::ltrue); continue;
+            case '|': result.push_back(logic::lor);   continue;
+            case '&': result.push_back(logic::land);  continue;
+            case '~': result.push_back(logic::lnot);  continue;
+                // clang-format on
+        }
+        CELER_VALIDATE(v == ' ',
+                       << "unexpected token '" << v
+                       << "' while parsing logic string");
+    }
+    if (reading_surf)
+    {
+        result.push_back(surf_id);
+    }
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a postfix logic expression to an infix expression.
+ *
+ * This is a helper class for \c convert_to_infix. It provides helper
+ * functions for building the infix expression using a stack.
+ */
+class ExprStack
+{
+  public:
+    //! A helper class for keeping track of the operand type of a
+    //! sub-expression.
+    struct Operand
+    {
+        logic::OperatorToken expr_type;
+        std::vector<logic_int> expr;
+    };
+
+    //! Push a binary operator.
+    void push_binary(logic_int op)
+    {
+        CELER_EXPECT(infix.size() > 1);
+        auto& op_2 = infix.back();
+        auto& op_1 = *(infix.end() - 2);
+        std::vector<logic_int> new_expr;
+        constexpr int max_extra_tokens = 5;
+        new_expr.reserve(max_extra_tokens + op_1.expr.size()
+                         + op_2.expr.size());
+        auto opposite = op == logic::lor ? logic::land : logic::lor;
+        this->add_sub_expr(new_expr, op_1.expr, opposite == op_1.expr_type);
+        new_expr.push_back(op);
+        this->add_sub_expr(new_expr, op_2.expr, opposite == op_2.expr_type);
+        infix.pop_back();
+        infix.pop_back();
+        infix.push_back({logic::OperatorToken{op}, new_expr});
+    }
+
+    //! Push an operand.
+    void push_unary(logic_int op)
+    {
+        CELER_EXPECT(!infix.empty());
+        auto&& [expr_type, expr] = infix.back();
+        std::vector<logic_int> new_expr;
+        constexpr int max_extra_tokens = 3;
+        new_expr.reserve(max_extra_tokens + expr.size());
+
+        new_expr.push_back(op);
+        this->add_sub_expr(new_expr, expr, expr_type < logic::lnot);
+        infix.pop_back();
+        infix.push_back({logic::OperatorToken{op}, new_expr});
+    }
+
+    //! Push a primitive (surface).
+    void push_primitive(logic_int elem)
+    {
+        // hijack ltrue as the token to represent a primitive
+        infix.push_back({logic::ltrue, {elem}});
+    }
+
+    //! Get the infix expression.
+    std::vector<logic_int> const& get_infix() const
+    {
+        CELER_EXPECT(infix.size() == 1);
+        return infix.front().expr;
+    }
```

Comment:
Either add or change:
```suggestion
    std::vector<logic_int> get_infix() &&
    {
        CELER_EXPECT(infix.size() == 1);
        return std::move(infix.front().expr);
    }
```